### PR TITLE
Improve protocol return types

### PR DESCRIPTION
**Pull request by srittau**
* Dunder comparisons must return bool.
* write() return type should be ignored.

### CODE REVIEW

**Comment by AlexWaygood** [comment]
FWIW, the reason I annotated the dunder comparison functions as returning `Any` rather than `bool` was @JelleZijlstra's comment [here](https://github.com/python/typeshed/pull/6449#issuecomment-982161219). But I don't have strong opinions about this :)

**Comment by srittau** [comment]
> Returning a non-bool from these methods isn't all that uncommon; numpy arrays do it.

I believe that's the case when the comparison operator are used for non-comparison purposes? For example, like pathlib uses the `/` operator for path concatenation? In that case I would suggest to use a custom protocol, instead of our typeshed protocol.

**Comment by AlexWaygood** [comment]
> > Returning a non-bool from these methods isn't all that uncommon; numpy arrays do it.
> 
> I believe that's the case when the comparison operator are used for non-comparison purposes?

I don't _think_ so; I think the issue is that the runtime doesn't care what these methods return, as long as it's something that `bool()` can be called on (which is _nearly_ everything). The most common approach is obviously to return `bool`; but that's not always the case, as the `pandas` example shows.

**Comment by srittau** [comment]
I feel that this exactly one of the cases that type checkers should catch. While of course everything is "boolean-like", returning arbitrary non-bools from those dunders seems error-prone. I am not making sense of the typing of numpy, but [the documentation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.__bool__.html) for `ndarray.__bool__` says: `self != 0`, so it seems it should return a `bool`.

**Comment by srittau** [comment]
Also, it's just one primer hit, despite the length of the output indicating something different.

**Comment by AlexWaygood** [comment]
> I feel that this exactly one of the cases that type checkers should catch. While of course everything is "boolean-like", returning arbitrary non-bools from those dunders seems error-prone. I am not making sense of the typing of numpy, but [the documentation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.__bool__.html) for `ndarray.__bool__` says: `self != 0`, so it seems it should return a `bool`.

Makes sense!

**Comment by Akuli** [comment]
`self != 0` doesn't return a bool in numpy:

```
>>> import numpy as np
>>> a = np.array([0,1,2,3])
>>> a != 0
array([False,  True,  True,  True])
>>> if a != 0: print("lol")
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Comment by JelleZijlstra** [comment]
Yes, in numpy all array operations do elementwise work where possible, so `x OP y` is broadly equivalent to `[a OP b for a, b in zip(x, y)]`. And numpy is very widely used, so I'm hesitant to make a change that drops some usage of numpy arrays.

However, we use these protocols both for functions like `min()` that really do need a bool (or something bool-like), and for the `operator.*` family of functions where the return value doesn't matter. We could perhaps split those up.

**Comment by srittau** [comment]
Splitting them sounds like a good idea. It would make most sense to me to keep the protocols in the PR as they are in the PR, and add custom protocols to `operator.pyi` directly.

**Comment by srittau** [comment]
Primer looks good now.

**Review by JelleZijlstra — APPROVED** [review]

File: stdlib/_operator.pyi [hunk 1]
```diff
@@ -25,6 +24,24 @@ _K = TypeVar("_K")
...
+class _SupportsDunderGT(Protocol):
+    def __gt__(self, __other: Any) -> Any: ...
+
+class _SupportsDunderLE(Protocol):
+    def __le__(self, __other: Any) -> Any: ...
+
+class _SupportsDunderGE(Protocol):
+    def __ge__(self, __other: Any) -> Any: ...
+
+_SupportsAnyComparison = _SupportsDunderLE | _SupportsDunderGE | _SupportsDunderGT | _SupportsDunderLT
```
Comment (AlexWaygood) [inline stdlib/_operator.pyi:43]: I've actually sort of come to regret the name I gave this Union in `_typeshed`. `SupportsAnyComparison` implies that "any comparison method will work" for an object satisfying this interface. But that's not what this interface actually represents.

`SupportsSomeComparison` might be a better name.

```suggestion
_SupportsSomeComparison = _SupportsDunderLE | _SupportsDunderGE | _SupportsDunderGT | _SupportsDunderLT
```

Comment (AlexWaygood) [reply stdlib/_operator.pyi:43]: Or just `_SupportsComparison`

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**